Lots of optimizations

forestrf opened the pull request:
I have done a lot of optimizations, large and small, without breaking saved mappings and configurations, and with slight API changes that should not require any change to the one using this library.

I have done some testing while working on it, step by step, and I did my best to make sure that there are no bugs, but with so many changes I can't be completely sure.

In build, here is the before and after (about 50% less CPU usage). I think it can be improved further.

| Starting point | This optimization |
| ------------- | ------------- |
| ![2018-11-18_21-35-07](https://user-images.githubusercontent.com/5580666/48677764-f2178d00-eb79-11e8-87d7-7e3bea160667.png) | ![2018-11-18_21-39-35](https://user-images.githubusercontent.com/5580666/48677796-8bdf3a00-eb7a-11e8-8b9a-b202f8445d63.png) |

The scene used for this test is Local Multiplayer with 2 characters (one in a gamepad and the other in keyboard+mouse)

Thank you.

forestrf committed:
Removed some garbage generation by caching some System.Enum.Parse results

forestrf committed:
Some optimizations + if GetKey is false, GetKeyDown will always also be false (added tests to verify it)

forestrf committed:
Some micro optimizations

forestrf committed:
Simplified signature of several functions + removed unusued comments of some functions

forestrf committed:
Small code reduction. Not an optimization

forestrf committed:
Reused array and changed a ToUpper with a string.Equals that ignores Case

forestrf committed:
connectedGamepads is always equal to _gamepads.Length. Changed connectedGamepads to a property to prevent external modifications
Merged two variables to a property of type { get; private set; }

forestrf committed:
Used [RuntimeInitializeOnLoadMethod] to initialize SInput automagically

forestrf committed:
Merged two variables to a property of type { get; private set; }

forestrf committed:
SInput.cs is not inside a namespace, generating conflicts with other classes that have the same name as those inside SinputSystems. Removed `using SinputSystems;` and manually added `SinputSystems.` to where is needed to prevent conflicts.

forestrf committed:
Removed static temp variables that are not needed + cached some calls to Enum.GetValues

forestrf committed:
Changed strings comparisons with int comparisons by pre hasing strings and comparing hashes. The case of two control names having the same hash is nearly impossible
Some micro optimizations

forestrf committed:
Cached String.Format for axis

forestrf committed:
Small changes

forestrf committed:
PrefersDeltaUse speedup by calculating inside Control/SmartControl during SInput.SinputUpdate();

forestrf committed:
Optimizations: Button states uses less logic code, and removed button checks from ShootyPlayer when checking it doesn't matter

forestrf committed:
Small optimizations

forestrf committed:
Small changes (for => foreach)

SophieH commented:
It will probably take me a while before I can get round to reviewing this properly but a quick scan tells me this is some great work! I'm impressed you managed to figure out so much of the system and improve so much in such a short time! I hope to merge this into the main branch before I get back to work on it :)

Thanks so much!

forestrf committed:
Bug fix (gamepad 7 tested two times, gamepad 8 not tested)
Bug fix (While searching for the highest axis to use it, a positive axis could override a negative axis of greater magnitude)
Removed axisAsButtonHeld from ControlState as it servers the same purpose as Held in all cases to make the code slightly easier
Small micro optimizations

forestrf committed:
Reused cached array 2 more times

forestrf committed:
Optimization: Removed extra button state check by reusing the AxisCheck
Optimization: Replaced some Input.GetKey to use a KeyCode instead of a String

forestrf commented:
So I made another optimization, but this time I only saved 0.05 ms. I found that ButtonHeldCheck is like a downgraded version of AxisCheck, that was already executed, but faster if you only want to check keys instead of axis. But given that we always call AxisCheck and that the button status can be obtained from it, ButtonHeldCheck can be removed.

It is getting difficult. I would like to stabiliza the graph though.

| Previous optimization  | This optimization |
| ------------- | ------------- |
| ![2018-11-19_22-37-54](https://user-images.githubusercontent.com/5580666/48736515-c5cc4100-ec4b-11e8-8e93-f858c2e77dc2.png)  | ![2018-11-19_22-36-17](https://user-images.githubusercontent.com/5580666/48736520-c82e9b00-ec4b-11e8-8853-9ec159f15db2.png)  |

forestrf committed:
Initialization allowed manually. Unity initializes it automatically thanks to the attribute, but it is useful to be able to initialize it manually too

forestrf committed:
Micro optimization and removed casting

forestrf committed:
Fix input index instead of a gamead slot.
